src/Build/BackEnd/Components/RequestBuilder/IntrinsicTasks/ItemGroupIntrinsicTask.cs

```diff
@@ -424,47 +424,44 @@ ISet<string> removeMetadata
                 }
                 else
                 {
-                    try
-                    {
-                        // The expression is not of the form "@(X)". Treat as string
-
-                        // Pass the non wildcard expanded excludes here to fix https://github.com/Microsoft/msbuild/issues/2621
-                        (string[] includeSplitFiles, FileMatcher.SearchAction action, string logExcludeSpec) = EngineFileUtilities.GetFileListEscaped(
-                            Project.Directory,
-                            includeSplit,
-                            excludes);
+                    // The expression is not of the form "@(X)". Treat as string
 
-                        if (action == FileMatcher.SearchAction.ReturnLogDriveEnumerationWildcard)
-                        {
-                            if (string.IsNullOrWhiteSpace(logExcludeSpec))
-                            {
-                                LoggingContext.LogWarning("WildcardResultsInDriveEnumeration", EscapingUtilities.UnescapeAll(includeSplit), XMakeAttributes.include, XMakeElements.itemGroup);
-                            }
-                        }
+                    // Pass the non wildcard expanded excludes here to fix https://github.com/dotnet/msbuild/issues/2621
+                    var getFilesResults = EngineFileUtilities.GetFileListEscaped(
+                        Project.Directory,
+                        includeSplit,
+                        excludes);
 
-                        foreach (string includeSplitFile in includeSplitFiles)
+                    if (getFilesResults.Action == FileMatcher.SearchAction.LogDriveEnumeratingWildcard)
+                    {
+                        if (!getFilesResults.IsDriveEnumeratingExcludePattern)
```

Comment:
`Recent change:`

1. In many codepaths, including this one, I log a warning or error based on whether a drive enumerating pattern is detected. The reason it's hard to unify is because there are so many different loggers and logging contexts used (that are not unified). Is there a way to actually unify all of this so I can just log something within EngineFileUtilities instead of logging warnings and errors separately in 4-5 different codepaths?

Reply:
This is an excellent instinct but I'm not seeing one; `LoggingContext`, `ILoggingService`, and `IBuildEngine` are unfortunately pretty different. I suppose we could pass an `object LogThing` into `GetFileListEscaped` and do the three implementations in a switch there to at least keep them close together. That . . . may actually be an ok idea? What do you think?

Specifically, I think a question to ask here is: "if we did that, would we need the `GetFilesImplementationResults` struct, or could it simplify to the existing calling pattern and hide all the complexity inside `GetFileListEscaped`?" If it can hide the complexity and also reduce duplication that seems good.